src/view/treewalker.js

```diff
@@ -0,0 +1,448 @@
+		// Text is a specific parent, because contains string instead of childs.
+		// We decided to not enter to the Text except situations when walker is iterating over every single character,
+		// or the bound starts/ends inside the Text. So when the position is at the beginning or at the end of the Text
+		// we move it just before or just after Text.
+		if ( item instanceof TextProxy ) {
+			// Position is at the end of Text.
+			if ( item._index + item._data.length == item._textNode._data.length ) {
+				if ( this.direction == 'FORWARD' ) {
+					nextPosition = Position.createAfter( item._textNode );
+					// When we change nextPosition of returned value we need also update walker current position.
+					this.position = nextPosition;
+				} else {
+					previousPosition = Position.createAfter( item._textNode );
+				}
+			}
+
+			// Position is at the begining ot the text.
+			if ( item._index === 0 ) {
+				if ( this.direction == 'FORWARD' ) {
+					previousPosition = Position.createBefore( item._textNode );
+				} else {
+					nextPosition = Position.createBefore( item._textNode );
+					// When we change nextPosition of returned value we need also update walker current position.
+					this.position = nextPosition;
+				}
+			}
+		}
+
+		return {
+			done: false,
+			value: {
+				type: type,
+				item: item,
+				previousPosition: previousPosition,
+				nextPosition: nextPosition,
+				length: length
+			}
+		};
+	}
+}
+
+/**
+ * Type of the step made by {@link engine.view.TreeWalker}.
+ * Possible values: `'ELEMENT_START'` if walker is at the beginning of a node, `'ELEMENT_END'` if walker is at the end
+ * of node, or `'TEXT'` if walker traversed over single and multiple characters.
+ *
+ * @typedef {String} engine.view.TreeWalkerValueType
+ */
+
+/**
+ * Object returned by {@link engine.view.TreeWalker} when traversing tree view.
+ *
+ * @typedef {Object} engine.view.TreeWalkerValue
+ * @property {engine.view.TreeWalkerValueType} type
+ * @property {engine.view.Item} item Item between old and new positions of {@link engine.view.TreeWalker}.
+ * @property {engine.view.Position} previousPosition Previous position of the iterator.
+ * * Forward iteration: For `'ELEMENT_END'` it is the last position inside the element. For all other types it is the
+ * position before the item. Note that it is more efficient to use this position then calculate the position before
+ * the node using {@link engine.view.Position.createBefore}.
+ * * Backward iteration: For `'ELEMENT_START'` it is the first position inside the element. For all other types it is
+ * the position after item.
+ * @property {engine.view.Position} nextPosition Next position of the iterator.
+ * * Forward iteration: For `'ELEMENT_START'` it is the first position inside the element. For all other types it is
+ * the position after the item.
+ * * Backward iteration: For `'ELEMENT_END'` it is last position inside element. For all other types it is the position
+ * before the item.
```

Comment:
I miss the information that the position is always moved from the inside of the text node to its parent if the position is at the beginning or the end of the text node.